src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.SyncNamespace
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.SyncNamespace), Shared]
+    internal sealed class CSharpSyncNamespaceCodeRefactoringProvider :
+        AbstractSyncNamespaceCodeRefactoringProvider<CSharpSyncNamespaceCodeRefactoringProvider, NamespaceDeclarationSyntax, CompilationUnitSyntax>
+    {             
+
+        protected override NamespaceDeclarationSyntax ChangeNamespace(NamespaceDeclarationSyntax node, ImmutableArray<string> namespaceParts)
+        {
+            return node.WithName(CreateNameSyntax(namespaceParts, aliasQualifier: null, namespaceParts.Length - 1).WithTriviaFrom(node.Name));
+        }  
+
+        protected override ImmutableArray<ISymbol> GetDeclaredSymbols(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            SyntaxList<MemberDeclarationSyntax> declarations;
+
+            if (node is NamespaceDeclarationSyntax namespaceDecl)
+            {
+                declarations = namespaceDecl.Members;
+            }
+            else if (node is CompilationUnitSyntax compilationUnit)
+            {
+                declarations = compilationUnit.Members;
+            }
+            else
+            {
+                return default;
+            }
+
+            var builder = ArrayBuilder<ISymbol>.GetInstance();
+            foreach (var declaration in declarations)
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(declaration, cancellationToken);
+                builder.AddIfNotNull(symbol);
+            }
+            return builder.ToImmutableAndFree();
+        }
+
+        private static NameSyntax GetQualifiedNameSyntax(NameSyntax node)
```

Comment:
this is basically just walking up the right side of a QualifiedName right?  There's no need to walk up multiple levels for that.  just:

```c#
if (node.Parent is QualifiedNameSyntax qualName && qualName.Right == node)
    node = qualName;
```